README.md

```diff
@@ -2,11 +2,41 @@
 
 [![CircleCI](https://circleci.com/gh/pkamenarsky/concur-replica.svg?style=svg)](https://circleci.com/gh/pkamenarsky/concur-replica)
 
-`concur-replica` is a [Concur](https://github.com/ajnsit/concur) backend for [Replica](https://github.com/pkamenarsky/replica). As such, you might want to make yourself familiar with both projects first.
+Build interactive web UIs using nothing but Haskell.
+
+No user-written JS. No transpiler.
+
+## How?!
+
+[Replica](https://github.com/pkamenarsky/replica): a remote virtual DOM library. Clients run a fixed JavaScript snippet which opens a connection to your Haskell server. The client uses it to notify the server of browser events, and the server uses it to send DOM updates to be applied by the client.
+
+[Concur](https://github.com/ajnsit/concur): a UI framework that generalizes the Elm architecture. It's based around `Widget`s, sections of the UI that may return a value.
+
+## Upsides and downsides
+
+This strategy is well understood due to its use by [Phoenix LiveView](https://hexdocs.pm/phoenix_live_view) (among other projects).
+
+There are two situations where you should rule `concur-replica` out completely:
+
++ Instant feedback is required
```

Comment:
I'd qualify that a bit - the feedback is more or less dependent on the ping between server and client. Also, in many situations (i.e. clicking on a button), it's pretty hard to tell the difference between 1ms and 50ms. Typing is another matter of course (i.e. client side prediction in replica, etc).

Maybe clarify a bit & move to the "softer costs" section?

Reply:
> Also, in many situations (i.e. clicking on a button), it's pretty hard to tell the difference between 1ms and 50ms.

Totally! A lot of the time I don't notice the lag with this strategy. However, internet is often choppy and you'll have spikes of more serious lag. I guess this kind of blends into the second point-- a good offline experience can mean "works entirely offline", but can also mean "you don't notice being offline for 1s".

However, I still think this deserves its own bullet point, because developers think about those as two different situations. They may not need Google Maps to work offline, but they still want it to be a good experience with 100ms lag. So whatever bullet point goes here should be the description of why efficient Google Maps is out. Thoughts?

Reply:
Put more clearly: do we have a one sentence way of saying why efficient Google Maps is out?

Reply:
Yep, agree!

How about "No client side logic is required"? I'd say that disqualifies Google Maps but also *doesn't* more traditional SPAs with lots of buttons and other such comparable elements.

In general I'd say, if Rails (without JS) can do it, Replica can as well. So maybe flip it around and say "Server side logic is sufficient" or something along those lines?

Reply:
I think the preceding "How?!" section already describes that concur-replica is a server-side only framework, so saying it again here isn't necessary (if that isn't clear we should expand the "How?!" section).

What I'd like to convey here is the consequences of that.

The streaming examples I actually see as a "soft" cost. The Phoenix LiveView people do use it for animations, eg this https://elixir-doom-fire.herokuapp.com/ (slightly lol example) Whether or not you'd use concur-replica for an app with animations depends-- if it's a personal blog showing how to use a haskell animation library it might be a great choice, if it's meant to have millions of users it's a bad one.

That's why I see the bandwidth issue as a "soft" cost, a situational one.

However, the instant feedback issue is a hard one. No matter how big a server you buy, using concur-replica instead of client side JS will always increase the amount of time until the user gets feedback. Even if you're willing to pay a high server and bandwidth bill, you can't get around this issue. That's why I see it as the "hard" cost.

Additional thoughts:

1. I added another commit mentioning in the section below that both events AND updates cost bandwidth, and moving the mention of "animations" into a parenthetical example.

2. I'm finding this conversation useful beyond just this PR, because I'm learning more about this strategy of doing a UI.

3. I've probably reached the depth of my knowledge on this issue, and won't have too much more to add. Want to take over this PR, reword it to your satisfaction, and then merge it with my blessing? We can always come back to it later once we've had more time to think about it. Also, if you do this I think the line *"if Rails (without JS) can do it, Replica can as well"* might be worth trying to work in, I think that summarizes things well.

Reply:
Good points, and I agree.

I did some minor adjustments, please look them over when you have the time and I'll merge 👍

Reply:
I'm :100: behind those changes, go for it!